tests/gen/gen_auto.go

```diff
@@ -0,0 +1,245 @@
+	DefaultStrLen = 10
+	// DefaultIntMin is the default minimum value of an integer to generate.
+	DefaultIntMin = 0
+	// DefaultIntMax is the default maximum value of an integer to generate.
+	DefaultIntMax = 10000
+)
+
+// AutoGenerateFromSDL generates random documents from a schema.
+func AutoGenerateFromSDL(gqlSDL string, options ...Option) ([]GeneratedDoc, error) {
+	genConfigs, err := parseConfig(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	typeDefs, err := parseSchema(gqlSDL)
+	if err != nil {
+		return nil, err
+	}
+	generator := newRandomDocGenerator(typeDefs, genConfigs)
+	return generator.generateDocs(options...)
+}
+
+// AutoGenerate generates random documents from collection definitions.
+func AutoGenerate(definitions []client.CollectionDefinition, options ...Option) ([]GeneratedDoc, error) {
+	err := validateDefinitions(definitions)
+	if err != nil {
+		return nil, err
+	}
+	typeDefs := make(map[string]client.CollectionDefinition)
+	for _, def := range definitions {
+		typeDefs[def.Description.Name] = def
+	}
+	generator := newRandomDocGenerator(typeDefs, nil)
+	return generator.generateDocs(options...)
+}
+
+func newRandomDocGenerator(types map[string]client.CollectionDefinition, config configsMap) *randomDocGenerator {
+	if config == nil {
+		config = make(configsMap)
+	}
+	configurator := newDocGenConfigurator(types, config)
+	return &randomDocGenerator{
+		configurator:     configurator,
+		generatedDocKeys: make(map[string][]string),
+	}
+}
+
+type randomDocGenerator struct {
+	configurator docsGenConfigurator
+
+	generatedDocKeys map[string][]string
+	random           rand.Rand
+}
+
+func (g *randomDocGenerator) generateDocs(options ...Option) ([]GeneratedDoc, error) {
+	err := g.configurator.Configure(options...)
+	if err != nil {
+		return nil, err
+	}
+
+	g.random = *g.configurator.random
+
+	resultDocs := make([]GeneratedDoc, 0, g.getMaxTotalDemand())
+	docsLists, err := g.generateRandomDocs(g.configurator.typesOrder)
+	if err != nil {
+		return nil, err
+	}
+	for _, docsList := range docsLists {
+		typeDef := g.configurator.types[docsList.ColName]
+		for _, doc := range docsList.Docs {
+			resultDocs = append(resultDocs, GeneratedDoc{
+				ColName: typeDef.Description.Name,
+				JSON:    createDocJSON(&typeDef, doc),
+			})
+		}
+	}
+	return resultDocs, nil
+}
+
+func (g *randomDocGenerator) getMaxTotalDemand() int {
+	totalDemand := 0
+	for _, demand := range g.configurator.docsDemand {
+		totalDemand += demand.max
+	}
+	return totalDemand
+}
+
+// getNextPrimaryDocKey returns the key of the next primary document to be used as a relation.
+func (g *randomDocGenerator) getNextPrimaryDocKey(secondaryType string, field *client.FieldDescription) string {
+	ind := g.configurator.usageCounter.getNextTypeIndForField(secondaryType, field)
+	docKey := g.generatedDocKeys[field.Schema][ind]
+	return docKey
+}
+
+func (g *randomDocGenerator) getDocKey(typeDef *client.CollectionDefinition, doc map[string]any) (string, error) {
```

Comment:
todo: `getFoo` suggests quite a cheap operation, typically a key lookup or something, but this function is actually doing an awful lot of work, serializing to JSON, then deserializing to `client.Document`, then getting the key.

If this function remains as such, I'd suggest renaming it to `generateDocKey` or similar.

I would however instead recommend a rework - it feels very wasteful to do all this work, only to then redo the json serialisation later (`GeneratedDoc` returns JSON atm), and then have that json deserialized (again) to `client.Document` by the integration test suite when these auto generated documents are consumed.

I don't think now is the appropriate place to look at returning `client.Documents`, as that would likely involve changing the core test suite stuff, but we could at least remove the need to serialise the document twice to json.

Reply:
refactored